### ISSUE TITLE
If you have two dimensions in your table + chart, Lightdash only allows you to visualize one metric

### ISSUE THREAD

**Issue by TuringLovesDeathMetal** [lightdash/lightdash#487]
![image](https://user-images.githubusercontent.com/31848148/134905072-d99ce157-c860-44b8-931d-364521d4165c.png)

I can't toggle on multiple metrics/table calculations into my chart in Lightdash. It only lets me toggle on one of the metrics, then the others are greyed out and won't untoggle. [See chart here](https://lightdash-analytics.lightdash.com/projects/21eef0b9-5bae-40f3-851e-9554588e71a6/tables/organizations_daily?dimensions=organizations_daily_days_since_organization_created%2Corganizations_daily_is_latest_organization&metrics=organizations_daily_count_unique_7d_active_organization_ids%2Corganizations_daily_count_unique_organization_ids&sort=%5B%7B%22fieldId%22%3A%22organizations_daily_is_latest_organization%22%2C%22descending%22%3Atrue%7D%2C%7B%22fieldId%22%3A%22organizations_daily_days_since_organization_created%22%2C%22descending%22%3Afalse%7D%5D&limit=500&column_order=organizations_daily_is_latest_organization%2Corganizations_daily_days_since_organization_created%2Corganizations_daily_count_unique_7d_active_organization_ids%2Corganizations_daily_count_unique_organization_ids%2Cweekly_active_organizations&selected_table_calculations=weekly_active_organizations&table_calculations=%5B%7B%22name%22%3A%22weekly_active_organizations%22%2C%22displayName%22%3A%22%25+weekly+active+organizations%22%2C%22sql%22%3A%22%28%24%7Borganizations_daily.count_unique_7d_active_organization_ids%7D%2F%24%7Borganizations_daily.count_unique_organization_ids%7D%29+*+100%22%7D%5D).

I'm going to try to make this bug happen in the Jaffle Shop demo and will update this ticket if I can!

**Comment by TuringLovesDeathMetal** [comment]
Yep, okay - here's teh same thing[ in the Jaffle shop demo](https://demo.lightdash.com/projects/62ead644-a616-4201-95e1-0ec424e09da1/tables/orders?dimensions=orders_status%2Corders_order_date&metrics=orders_unique_order_count%2Ccustomers_unique_customer_count&limit=500&column_order=orders_order_date%2Corders_status%2Corders_unique_order_count%2Ccustomers_unique_customer_count%2Ctest_calc_customer_count_order_count&selected_table_calculations=test_calc_customer_count_order_count&table_calculations=%5B%7B%22name%22%3A%22test_calc_customer_count_order_count%22%2C%22displayName%22%3A%22%25+test+calc+-+customer+count+%2F+order+count%22%2C%22sql%22%3A%22%24%7Bcustomers.unique_customer_count%7D%2F%24%7Borders.unique_order_count%7D%22%7D%5D)

Basically, if you have **two dimensions** (so you've grouped by one of them), then you can only visualize one metric.

**Comment by TuringLovesDeathMetal** [comment]
If you remove one of the dimensions from the visualization, then you ca have multiple metrics in the chart ([see example here](https://demo.lightdash.com/projects/62ead644-a616-4201-95e1-0ec424e09da1/tables/orders?dimensions=orders_status%2Corders_order_date&metrics=orders_unique_order_count%2Ccustomers_unique_customer_count&limit=500&column_order=orders_order_date%2Corders_status%2Corders_unique_order_count%2Ccustomers_unique_customer_count%2Ctest_calc_customer_count_order_count&selected_table_calculations=test_calc_customer_count_order_count&table_calculations=%5B%7B%22name%22%3A%22test_calc_customer_count_order_count%22%2C%22displayName%22%3A%22%25+test+calc+-+customer+count+%2F+order+count%22%2C%22sql%22%3A%22%24%7Bcustomers.unique_customer_count%7D%2F%24%7Borders.unique_order_count%7D%22%7D%5D))

**Comment by ZeRego** [comment]
@TuringLovesDeathMetal so if we want to group by dimension and display multiple metrics, how would the chart look like? 

When we group by dimension: 
x axis: Order status
y axis: Orders unique order count
top labels: all the dates  e.g 2018-03-18
  
When we have multiple metrics: 
x axis: Order status
y axis: no label
top labels: all the metrics

When we group by dimension and have multiple metrics: 
x axis: Order status
y axis: no label
top labels: a combination between each grouped value and metric type ?

e.g 
2018-03-18 Order unique order count
2018-03-18 Order unique customer count
2018-03-19 Order unique order count
2018-03-19 Order unique customer count

**Comment by owlas** [comment]
This was actually a design choice. It get's pretty wild to visualise 4 dimensions at once. Usually 3 is pretty easy:

* y-axis
* x-axis
* hue / colour / line per group

For 4 dimensions it gets pretty tricky. Bar charts can have grouped bars, not so bad. But what would the line chart look like ideally @TuringLovesDeathMetal in your example?

**Comment by ZeRego** [comment]
Have a line for each metric in a timeline. Example with dummy data:

![Uploading Screenshot 2021-09-29 at 14.38.04.png…]()

**Comment by TuringLovesDeathMetal** [comment]
^yep, exactly.

**Comment by TuringLovesDeathMetal** [comment]
Example of this:
- product usage for user cohorts (based on month of activation) over time for active vs. non-active users

**Comment by TuringLovesDeathMetal** [comment]
Example of this:
- product usage for user cohorts (based on month of activation) over time for active vs. non-active users

**Comment by TuringLovesDeathMetal** [comment]
Another example:
- % weekly active companies overall
- I have % of Weekly active companies older than 3 months
- % WAC younger than 3 months
- I want to see this split by cloud vs. self-hosted

I'd start with the question: 
"Is the engagement of companies decreasing over time because new companies are less engaged?"

If we found, yes, this was true. 

Then, I'd want to ask: "Is this happening in Cloud, self-hosted, or both?"

^it's a stupidly complicated graph, but I'm using it to continue to ask questions, figure out the underlying problem, then I'd break these "bits of the story" out into separate, clearer plots in an analysis.

**Comment by TuringLovesDeathMetal** [comment]
Sorry, one last example (just making my case here 😅 )

If you have a graph of % of weekly active users over time, then you probably also want "count of weekly active users"

^This is mostly for external reporting of things. E.g. 1 million weekly active users sounds way more impressive to people than "40% of users are weekly active". I've found that for C-suite/Exec meetings, they always wanted the raw numbers as well as %s on the same graphs so they could do the "mental maths" of costs, support needed if we launched a new product, etc.

**Comment by owlas** [comment]
> Agreed, it's chaotic - but you likely click off a bunch of the lines. I've done these types of crazy wild charts in the past really just for trying to uncover trends. Not so much for presenting the final charts to end-users.

That makes sense, from a dataviz perspective it seems pretty unintuitive.

Could you get the same behaviour by making a derived column representing a new group label?

```sql
CONCAT(IF(${users.is_active}, 'Active ', ''), ${product.name}, ' user')
```

Would give:
| is_active | product.name | metric | label |
|---|---|---|---|
| 1 | pants|12|Active pants user|
|0|pants|93|pants user|
|1|socks|23|Active socks user|
|0|sockes|11|Socks user|

**Comment by TuringLovesDeathMetal** [comment]
hmmm, this looks like two dimensions and one metric 🤔 

What would you do in this case if you had two metrics and two dimensions? Are you saying that we shouldn't include the other metric? 

E.g. for this:
is_active | product.name | metric | label
-- | -- | -- | -- | --
1 | pants | 12 | 4% | Active pants user
0 | pants | 93 | 6% | pants user
1 | socks | 23 | 8% | Active socks user
0 | sockes | 11 | 5% | Socks user

**Comment by TuringLovesDeathMetal** [comment]
ohhhh I see what you're saying. You're talking about re-creating the x-axis to be a concatenated label. 

Hmmm, in the example you gave, that works. But in an example with a date-series, that would really break the visualization (I would want all of my lines to be plotted in the same date range, on top of each other - not broken apart)

e.g. plotting this:

date | product.name | metric1 | metric2 | label
-- | -- | -- | -- | --
01/02/03 | pants | 12 | 4% | 01/02/03 - pants
01/02/03 | socks | 93 | 6% | 01/02/03 - socks
02/02/03 | pants | 23 | 8% | 02/02/03 - pants
02/02/03 | socks | 11 | 5% | 02/02/03 - socks

**Comment by TuringLovesDeathMetal** [comment]
also, here's a thread of people complaining about this feature not existing in Looker 😅 

https://community.looker.com/dashboards-looks-7/measure-vs-measure-scatterplots-3-28-1296/index2.html


^I think this is quite similar. After the measure-by-measure scatterplots worked, they're asking for the same setup as we're talking about, but for scatterplots. i.e. you have two dimensions (or metrics) you're grouping your data by: one is represented as different colours, the other is the x-axis

**Comment by owlas** [comment]
That thread on the looker community is really interesting. It shows how if you give enough flexibility in the plots, you can go beyond just presenting stuff to actually doing exploration/analysis like checking correlations across groups.

This is awesome!

I've changed my mind, we should adopt this flexibility. We can keep users from making obvious mistakes by suggesting reasonable defaults (like only showing 1 metric + hue dimension, or `n` metrics and no hue dimension) but still allow configuration for more advanced types.

**Comment by TuringLovesDeathMetal** [comment]
+1 to having defaulted hues for a dimensions group. I'd say we take the first dimension in the group and base the hue grouping on that. 

So, for example, if I had the table above:

date | product.name | metric1 | metric2 | label
-- | -- | -- | -- | --
01/02/03 | pants | 12 | 4% | 01/02/03 - pants
01/02/03 | socks | 93 | 6% | 01/02/03 - socks
02/02/03 | pants | 23 | 8% | 02/02/03 - pants
02/02/03 | socks | 11 | 5% | 02/02/03 - socks

I'd want `01/02/03` to be red, and pants + socks to be a different shade of red
I'd want `02/02/03` to be blue, and pants + socks to be a different shade of blue

 (in an _**ideal**_ world, `pants` would always be the darkest shade, `socks` the lightest)

If we add 3x dimensions, we could introduce line styles. (e.g. dotted, dashed, etc.) to visualize the third grouping. 

4+ and you're just going to get different colours, because you're crazy

^This is like advanced version though (unless it's super easy to do!)

**Comment by owlas** [comment]
How about defaults? If I have 2 dimensions AND 2 metrics, should we default to showing everything or default to showing `1 metric + 1 x dimension + 1 group dimension` or `2 metrics + 1 x dimension` or all of it `2 metrics + 1 x dimension + 1 group dimension`?

**Comment by owlas** [comment]
That thread also made me realise we should allow metrics to appear on the x-axis, if desired.

**Comment by TuringLovesDeathMetal** [comment]
default to showing everything

**Comment by owlas** [comment]
It'd be great if we could have sensible defaults and unlimited flexibility. Like the sorts, we guess how you'd want it sorted but you can change it however you want and we never override your sorts if you set the manually.

Could be the same for the plot, e.g. we pick the dimensions + metrics that make sense (if you didn't set any) but you're always free to change them
